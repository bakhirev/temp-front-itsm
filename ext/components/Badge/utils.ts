import type { ITheme } from '../themes';

import type { Kind } from './BadgeComponent';

export const getBackgroundColor = (kind: Kind, theme: ITheme) => {
  switch (kind) {
    case 'info':
      return theme.color.primary[60];
    case 'success':
      return theme.color.success[50];
    case 'warning':
      return theme.color.warning[50];
    case 'error':
      return theme.color.error[60];
    case 'neutral-grey':
      return theme.color.neutral[50];
    case 'neutral-dark':
      return theme.color.neutral[80];
    case 'neutral-light':
    case 'neutral-light-disabled':
    case 'neutral-light-inactive':
      return theme.color.neutral[10];
    case 'neutral-white':
    case 'neutral-white-disabled':
    case 'neutral-white-inactive':
      return theme.color.neutral.white;
  }
};

export const getTextColor = (kind: Kind, theme: ITheme) => {
  switch (kind) {
    case 'neutral-light':
    case 'neutral-white':
      return theme.color.neutral[90];
    case 'neutral-light-disabled':
    case 'neutral-white-disabled':
      return theme.color.neutral[30];
    case 'neutral-light-inactive':
    case 'neutral-white-inactive':
      return theme.color.neutral[50];
    case 'success':
    case 'info':
    case 'warning':
    case 'error':
    case 'neutral-grey':
    case 'neutral-dark':
      return theme.color.neutral.white;
  }
};
