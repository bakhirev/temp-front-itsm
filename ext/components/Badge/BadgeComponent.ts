import styled, { css } from 'styled-components';

import { DEFAULT_THEME } from '../common';

import { getBackgroundColor, getTextColor } from './utils';

const BORDER_RADIUS_BIG = 10;
const BORDER_RADIUS_SMALL = 8;

const HEIGHT_BIG = 20;
const HEIGHT_SMALL = 16;

const HORIZONTAL_PADDING_BIG = 7;
const HORIZONTAL_PADDING_SMALL = 5;

const MIN_WIDTH_BIG = 22;
const MIN_WIDTH_SMALL = 17;

export type Kind =
  | 'neutral-grey'
  | 'neutral-dark'
  | 'info'
  | 'warning'
  | 'error'
  | 'success'
  | 'neutral-light'
  | 'neutral-light-disabled'
  | 'neutral-light-inactive'
  | 'neutral-white'
  | 'neutral-white-disabled'
  | 'neutral-white-inactive';

export type Size = 'big' | 'small';

interface IBadgeComponentProps {
  kind: Kind;
  size: Size;
}

export const BadgeComponent = styled.div<IBadgeComponentProps>`
  ${({ kind, size, theme }) => css`
    align-items: center;
    background-color: ${getBackgroundColor(kind, theme)};
    border-radius: ${size === 'big' ? BORDER_RADIUS_BIG : BORDER_RADIUS_SMALL}px;
    color: ${getTextColor(kind, theme)};
    display: flex;
    height: ${size === 'big' ? HEIGHT_BIG : HEIGHT_SMALL}px;
    justify-content: center;
    min-width: ${size === 'big' ? MIN_WIDTH_BIG : MIN_WIDTH_SMALL}px;
    padding: 0 ${size === 'big' ? HORIZONTAL_PADDING_BIG : HORIZONTAL_PADDING_SMALL}px;
    user-select: none;
  `};
`;

BadgeComponent.defaultProps = {
  theme: DEFAULT_THEME,
};
