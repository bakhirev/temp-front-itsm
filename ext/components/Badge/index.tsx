import React, { FC } from 'react';

import { Body2Long, Caption1 } from '../Typography';

import { BadgeComponent } from './BadgeComponent';
import type { Kind, Size } from './BadgeComponent';

export interface IBadgeProps {
  /** Имя класса для переопределения стилей */
  className?: string;
  /** data-test-id атрибут для тестирования компонента */
  dataTestId?: string;
  /** Тип компонента */
  kind?: Kind;
  /** Размер компонента */
  size?: Size;
  /** Значение, отображаемое в бейдже */
  value: number;
}

export const Badge: FC<IBadgeProps> = ({
  className,
  dataTestId,
  kind = 'neutral-light',
  size = 'big',
  value,
}) => {
  const Label = size === 'big' ? Body2Long : Caption1;
  return (
    <BadgeComponent className={className} data-test-id={dataTestId} kind={kind} size={size}>
      <Label>{value}</Label>
    </BadgeComponent>
  );
};
