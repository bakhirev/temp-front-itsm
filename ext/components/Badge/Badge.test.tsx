import React from 'react';
import { shallow } from 'enzyme';

import { Badge } from '../Badge';

jest.mock('../common', () => ({
  DEFAULT_THEME: {},
}));

describe('Badge', () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  const requiredProps = {
    value: 5,
  };

  it('should render component with required properties, kind NEUTRAL_LIGHT and big size by default', () => {
    const wrapper = shallow(<Badge {...requiredProps} />);
    expect(wrapper).toMatchSnapshot();
  });

  it('should render component with small size', () => {
    const wrapper = shallow(<Badge {...requiredProps} size="small" />);
    expect(wrapper).toMatchSnapshot();
  });

  it('should render component with kind INFO', () => {
    const wrapper = shallow(<Badge {...requiredProps} kind="info" />);
    expect(wrapper).toMatchSnapshot();
  });

  it('should render component with kind NEUTRAL_GREY', () => {
    const wrapper = shallow(<Badge {...requiredProps} kind="neutral-grey" />);
    expect(wrapper).toMatchSnapshot();
  });

  it('should render component with kind NEUTRAL_DARK', () => {
    const wrapper = shallow(<Badge {...requiredProps} kind="neutral-dark" />);
    expect(wrapper).toMatchSnapshot();
  });

  it('should render component with kind SUCCESS', () => {
    const wrapper = shallow(<Badge {...requiredProps} kind="success" />);
    expect(wrapper).toMatchSnapshot();
  });

  it('should render component with kind ERROR', () => {
    const wrapper = shallow(<Badge {...requiredProps} kind="error" />);
    expect(wrapper).toMatchSnapshot();
  });

  it('should render component with kind NEUTRAL_LIGHT_DISABLED', () => {
    const wrapper = shallow(<Badge {...requiredProps} kind="neutral-light-disabled" />);
    expect(wrapper).toMatchSnapshot();
  });

  it('should render component with kind NEUTRAL_LIGHT_INACTIVE', () => {
    const wrapper = shallow(<Badge {...requiredProps} kind="neutral-light-inactive" />);
    expect(wrapper).toMatchSnapshot();
  });

  it('should render component with kind NEUTRAL_WHITE', () => {
    const wrapper = shallow(<Badge {...requiredProps} kind="neutral-white" />);
    expect(wrapper).toMatchSnapshot();
  });

  it('should render component with kind NEUTRAL_WHITE_DISABLED', () => {
    const wrapper = shallow(<Badge {...requiredProps} kind="neutral-white-disabled" />);
    expect(wrapper).toMatchSnapshot();
  });

  it('should render component with kind NEUTRAL_WHITE_INACTIVE', () => {
    const wrapper = shallow(<Badge {...requiredProps} kind="neutral-white-inactive" />);
    expect(wrapper).toMatchSnapshot();
  });

  it('should render component with kind WARNING', () => {
    const wrapper = shallow(<Badge {...requiredProps} kind="warning" />);
    expect(wrapper).toMatchSnapshot();
  });
});
