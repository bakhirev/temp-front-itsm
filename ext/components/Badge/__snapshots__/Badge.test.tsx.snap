// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Badge should render component with kind ERROR 1`] = `
<styled.div
  kind="error"
  size="big"
  theme={Object {}}
>
  <styled.div>
    5
  </styled.div>
</styled.div>
`;

exports[`Badge should render component with kind INFO 1`] = `
<styled.div
  kind="info"
  size="big"
  theme={Object {}}
>
  <styled.div>
    5
  </styled.div>
</styled.div>
`;

exports[`Badge should render component with kind NEUTRAL_DARK 1`] = `
<styled.div
  kind="neutral-dark"
  size="big"
  theme={Object {}}
>
  <styled.div>
    5
  </styled.div>
</styled.div>
`;

exports[`Badge should render component with kind NEUTRAL_GREY 1`] = `
<styled.div
  kind="neutral-grey"
  size="big"
  theme={Object {}}
>
  <styled.div>
    5
  </styled.div>
</styled.div>
`;

exports[`Badge should render component with kind NEUTRAL_LIGHT_DISABLED 1`] = `
<styled.div
  kind="neutral-light-disabled"
  size="big"
  theme={Object {}}
>
  <styled.div>
    5
  </styled.div>
</styled.div>
`;

exports[`Badge should render component with kind NEUTRAL_LIGHT_INACTIVE 1`] = `
<styled.div
  kind="neutral-light-inactive"
  size="big"
  theme={Object {}}
>
  <styled.div>
    5
  </styled.div>
</styled.div>
`;

exports[`Badge should render component with kind NEUTRAL_WHITE 1`] = `
<styled.div
  kind="neutral-white"
  size="big"
  theme={Object {}}
>
  <styled.div>
    5
  </styled.div>
</styled.div>
`;

exports[`Badge should render component with kind NEUTRAL_WHITE_DISABLED 1`] = `
<styled.div
  kind="neutral-white-disabled"
  size="big"
  theme={Object {}}
>
  <styled.div>
    5
  </styled.div>
</styled.div>
`;

exports[`Badge should render component with kind NEUTRAL_WHITE_INACTIVE 1`] = `
<styled.div
  kind="neutral-white-inactive"
  size="big"
  theme={Object {}}
>
  <styled.div>
    5
  </styled.div>
</styled.div>
`;

exports[`Badge should render component with kind SUCCESS 1`] = `
<styled.div
  kind="success"
  size="big"
  theme={Object {}}
>
  <styled.div>
    5
  </styled.div>
</styled.div>
`;

exports[`Badge should render component with kind WARNING 1`] = `
<styled.div
  kind="warning"
  size="big"
  theme={Object {}}
>
  <styled.div>
    5
  </styled.div>
</styled.div>
`;

exports[`Badge should render component with required properties, kind NEUTRAL_LIGHT and big size by default 1`] = `
<styled.div
  kind="neutral-light"
  size="big"
  theme={Object {}}
>
  <styled.div>
    5
  </styled.div>
</styled.div>
`;

exports[`Badge should render component with small size 1`] = `
<styled.div
  kind="neutral-light"
  size="small"
  theme={Object {}}
>
  <styled.div>
    5
  </styled.div>
</styled.div>
`;
